packages/buidler-etherscan/src/index.ts

```diff
@@ -46,30 +47,23 @@ task("verify-contract", "Verifies contract on etherscan")
         );
       }
 
-      let source: string;
-      try {
-        source = await run(TASK_FLATTEN_GET_FLATTENED_SOURCE);
-      } catch (_) {
-        throw new BuidlerPluginError(
-          `Your ${taskArgs.contractName} contract constains a cyclic dependency, Etherscan doesn't currently support contracts with such dependencies through its API, please use their GUI at https://etherscan.io/verifyContract to verify your contract.`
-        );
-      }
-
       await run(TASK_COMPILE);
       const abi = (await readArtifact(
         config.paths.artifacts,
         taskArgs.contractName
       )).abi;
       config.solc.fullVersion = await getLongVersion(config.solc.version);
 
+      const source = JSON.stringify(await run(TASK_COMPILE_GET_COMPILER_INPUT));
+
       const request = toRequest({
         apiKey: etherscan.apiKey,
         contractAddress: taskArgs.address,
         sourceCode: source,
-        contractName: taskArgs.contractName,
+        contractName: `contracts/${taskArgs.contractName}.sol:${taskArgs.contractName}`,
```

Comment:
This line is the only problematic thing I can see, as it imposes the same names for files and contracts.

This could be fixed by doing something like

```ts
const solcInput : SolcInput = JSON.stringify(await run(TASK_COMPILE_GET_COMPILER_INPUT)); 

const filesWithContract = Object.entries(solcInput.sources).find((fileGlobalName, {content}) => hasContractDeclaration(content, taskArgs.contractName)).map((globalName) => globalName);

if (filesWithContract.length === 0) {
   // contract not found
   throw BuidlerPluginError(...);
}

if (filesWithContract.length > 1) {
   // throw error? what should be done here??
}

const contractName = `${filesWithContract[0}:${taskArgs.contractName}`;
```

Reply:
Yes, you're absolutely right, it's a known limitation for this PR.

My idea for handling any use case is extending the `verify-contract` task arguments to accept an optional `contractPath` argument (e.g. contracts/A.sol) to build the complete contract name as `${taskArgs.contractPath}:${taskArgs.contractName}`; when the `contractPath` is not given, the name `contracts/${taskArgs.contractName}.sol:${taskArgs.contractName}` will be used.

Another option could be changing the `contractName` argument of the `verify-contract` task to become the full contract name, specified exactly as expected by Etherscan API (e.g. contracts/A.sol:B.sol).

What do you think?

Reply:
@alcuadrado Does this issue gate merging? Thanks!

Reply:
@pcowgill this issue has been resolved in last commit

Reply:
@canepat Great, thanks. I thought so but I wanted to confirm that